specs/keccak256/variable-length-input.md

```diff
@@ -0,0 +1,147 @@
+# Keccak256 For Variable Length Input
+
+We describe the circuit that can process inputs of an arbitrary size.
+
+## Plain Keccak behavior
+
+Keccak256 uses a simple iterating sponge construction to handle a variable length input. The sponge absorbs 17 words(136 bytes) of inputs, applies the `Keccak-f` permutation, and then absorb next 17 words. Once all parts of the inputs are absorbed, the sponge squeezes out the 32 bytes for the output.
+
+If the last part of the input is less than 136 bytes, a padding must be applied. The padding is a multi-rate padding, it pads single bit 1 followed by minimum number of bits 0 followed by a single bit 1, such that the result is 136 bytes.
+
+## Circuit behavior
+
+### Lookup Region
+
+This region splits the input to multiple parts, each part corresponds to their `Keccak-f` permutation round.
+This is also a lookup table for the other circuits to lookup the Keccak256 input to the output.
+
+Define:
+
+We use `Round` for the cells of the following columns in a row, it represents the information in a `keccak_f` round.
+
+The `curr: Round` and `next: Round` represent the current and the next row.
+
+Columns:
+
+- `hash_id` Sequential number for grouping different hash
```

Comment:
I'm still not sure how can we have two `hash_id`s within the same circuit instance. In case we have it because all the permutations of all hashes will share some common region/table, couldn't we use the RLC(input) as id? I think makes a bit more sense as it ties the input to the permutation computations.

Reply:
Using input as the id might not make sense.
The definition of the input is "the RLC of the 136 bytes to be absorbed in this round". Let's say if we have two rows with the same input value 123,  their output could be different if one is the first 136 bytes of the acc input and the other is the second 136 bytes of the acc input.

Reply:
But within the same permutation input computation, we already enforce the continuity and the relation between each byte by constraining the RLC.

If the RLC matches, the bytes were for sure from the same ID no?